homeassistant/components/calendar/__init__.py

```diff
@@ -0,0 +1,260 @@
+        }
+
+    @property
+    def track(self):
+        """Are we tracking events on this search sensor."""
+        return self._track
+
+    @property
+    def is_all_day(self):
+        """If the next event is all day then True."""
+        return self._cal_data.get('all_day', False)
+
+    @property
+    def search(self):
+        """Return the search string."""
+        return self._search
+
+    @property
+    def name(self):
+        """Return the device name."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return if the event in process."""
+        return self._state
+
+    @property
+    def state(self):
+        """Return the state of the calendar event."""
+        return STATE_ON if self.is_on else STATE_OFF
+
+    def cleanup(self):
+        """Cleanup any start/end listeners that were setup."""
+        from homeassistant.const import EVENT_TIME_CHANGED
+        self._cal_data = {
+            'all_day': False,
+            'offset_time': 0,
+            'message': '',
+            'start': None,
+            'end': None,
+            'location': None,
+            'description': None
+        }
+
+        if self._start_listener:
+            self.hass.bus.remove_listener(EVENT_TIME_CHANGED,
+                                          self._start_listener)
+            self._start_listener = None
+        if self._end_listener:
+            self.hass.bus.remove_listener(EVENT_TIME_CHANGED,
+                                          self._end_listener)
+            self._end_listener = None
+
+    def get_next_event(self):
+        """Raise an error so that subclasses implement this."""
+        raise NotImplementedError()
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Search for the next event."""
+        from homeassistant.const import EVENT_TIME_CHANGED
+        from homeassistant.helpers.event import track_point_in_time
+
+        event = self.get_next_event()
+        if not event:
+            self.cleanup()
+            return
+
+        def _get_date(date):
+            """Get the dateTime from date or dateTime as a local."""
+            if 'date' in date:
+                return dt.as_local(dt.dt.datetime.combine(
+                    dt.parse_date(date['date']), dt.dt.time()))
+            else:
+                return dt.parse_datetime(date['dateTime'])
+
+        self._cal_data['all_day'] = 'date' in event['start']
+        start = _get_date(event['start'])
+        end = _get_date(event['end'])
+
+        # check if the start/end times changed, if not then exit
+        if start == self._cal_data['start'] and end == self._cal_data['end']:
+            # don't redo the listener if we have the same start/end times
+            return
+
+        summary = event['summary']
+
+        # check if we have an offset tag in the message
+        import re
+        search = re.search('{}([0-9]+)'.format(self._offset), summary)
```

Comment:
I don't want to suggest even more scope creep but if you use voluptuous for config validation, you can use `config_validation.time_period` which supports values `+01:00`, `-00:00:05`, `{ 'hours': 5, 'minutes': 2}` and converts them all to `timedelta`